lib/redis_cluster/client.rb

```diff
@@ -22,29 +30,48 @@ def initialize(startup_hosts, configs = {})
     end
 
     def execute(method, args, &block)
-      ttl = Configuration::REQUEST_TTL
       asking = false
+      last_error = nil
+      retries_left = @retry_count
       try_random_node = false
 
-      while ttl > 0
-        ttl -= 1
+      # We use `>= 0` instead of `> 0` because we decrement this counter on the
+      # first run.
+      while retries_left >= 0
+        retries_left -= 1
+
         begin
           return @pool.execute(method, args, {asking: asking, random_node: try_random_node}, &block)
-        rescue Errno::ECONNREFUSED, Redis::TimeoutError, Redis::CannotConnectError, Errno::EACCES
+
+        rescue Errno::ECONNREFUSED, Redis::TimeoutError, Redis::CannotConnectError, Errno::EACCES => e
+          last_error = e
+
+          # Getting an error while executing may be an indication that we've
+          # lost the node that we were talking to and in that case it makes
+          # sense to try a different node and maybe reload our node pool (if
+          # the new node issues a `MOVE`).
           try_random_node = true
-          sleep 0.1 if ttl < Configuration::REQUEST_TTL / 2
+
         rescue => e
+          last_error = e
+
           err_code = e.to_s.split.first
           raise e unless %w(MOVED ASK).include?(err_code)
 
           if err_code == 'ASK'
             asking = true
           else
+            # `MOVED` indicates a permanent which means that our slot mappings
+            # our stale: reload them.
             reload_pool_nodes(false)
-            sleep 0.1 if ttl < Configuration::REQUEST_TTL / 2
           end
         end
       end
+
+      # If we ran out of retries (the maximum number may have been set to 0),
+      # surface any error that was thrown back to the caller. We'd otherwise
+      # suppress the error, which would return something quite unexpected.
+      raise last_error
```

Comment:
Behavior here has changed: previously, after hitting maximum TTL, the control loop would fail the condition `while ttl > 0` and this method would return `false`. The trouble with that behavior is that there's no way for the caller to differentiate that return value from an error, and that may cause it to do the wrong thing (in our stack, it accidentally started rate limiting people inappropriately for example).

Here we change the behavior so that if we do leave the loop unsuccessfully, we raise whatever error we got on the loop's last iteration to the caller so that it can recognize the error condition and has a chance to handle it.